File: src/snippets/graph/depth-first-search/depth-first-search.cs
using System;
using System.Collections.Generic;

class Graph
{
    private int Vertices;
    private List<int>[] adj;

    public Graph(int vertices)
    {
        Vertices = vertices;
        adj = new List<int>[vertices];
        for (int i = 0; i < vertices; i++)
            adj[i] = new List<int>();
    }

    public void AddEdge(int v, int w)
    {
        adj[v].Add(w);
    }

    private void DFSUtil(int v, bool[] visited)
    {
        visited[v] = true;
        Console.Write(v + " ");

        foreach (int neighbor in adj[v])
        {
            if (!visited[neighbor])
                DFSUtil(neighbor, visited);
        }
    }

    public void DFS(int v)
    {
        bool[] visited = new bool[Vertices];
        DFSUtil(v, visited);
    }

    public static void Main(string[] args)
    {
        Graph g = new Graph(4);

        g.AddEdge(0, 1);
        g.AddEdge(0, 2);
        g.AddEdge(1, 2);
        g.AddEdge(2, 0);
        g.AddEdge(2, 3);
        g.AddEdge(3, 3);

        Console.WriteLine("Depth First Traversal starting from vertex 2:");

        g.DFS(2);
    }
}

File: src/snippets/graph/depth-first-search/depth-first-search.lua
-- Depth First Search (DFS) implementation in Lua

-- Graph represented as an adjacency list
local graph = {
    A = {"B", "C"},
    B = {"D", "E"},
    C = {"F"},
    D = {},
    E = {"F"},
    F = {}
}

-- DFS function
local function DFS(graph, startNode, visited)
    if not visited then
        visited = {}
    end

    -- Mark the current node as visited
    visited[startNode] = true
    print(startNode)

    -- Recur for all the vertices adjacent to this vertex
    for _, neighbor in ipairs(graph[startNode]) do
        if not visited[neighbor] then
            DFS(graph, neighbor, visited)
        end
    end
end

-- Usage example
local startNode = "A"
print("Depth First Search starting from node " .. startNode .. ":")
DFS(graph, startNode)

-- Explanation:
-- The graph is represented as an adjacency list where each key is a node and the value is a list of adjacent nodes.
-- The DFS function takes the graph, a starting node, and an optional visited table to keep track of visited nodes.
-- It marks the current node as visited, prints it, and then recursively visits all unvisited adjacent nodes.